pkg/chunk/azure/blob_storage_client.go

```diff
@@ -110,13 +111,26 @@ func NewBlobStorage(cfg *BlobStorageConfig) (*BlobStorage, error) {
 func (b *BlobStorage) Stop() {}
 
 func (b *BlobStorage) GetObject(ctx context.Context, objectKey string) (io.ReadCloser, error) {
+	var cancel context.CancelFunc
 	if b.cfg.RequestTimeout > 0 {
-		// The context will be cancelled with the timeout or when the parent context is cancelled, whichever occurs first.
-		var cancel context.CancelFunc
 		ctx, cancel = context.WithTimeout(ctx, b.cfg.RequestTimeout)
-		defer cancel()
 	}
 
+	rc, err := b.getObject(ctx, objectKey)
```

Comment:
But rethinking about it: is even correct to apply the timeout to the download time (time it takes to download the whole response content)? We had a similar discussion in Thanos and we ended up it's very hard to set such timeouts (while we should circuit break on download speed below X bytes/sec).

Reply:
I also find it strange to have timeout applied at this level, but it's better than having it applied at multiple call sites.

> while we should circuit break on download speed below X bytes/sec

You could apply the same logic at many places, but generally we don't, and try anyway. Returning "timeout" before time really ran out would be quite interesting.

Reply:
> while we should circuit break on download speed below X bytes/sec
It would be harder to do and would be complex for users to understand I feel. I am also not sure how many users would use this vs setting a timeout for queries which would anyways cancel the context.